spec/integration/parser/parameter_defaults_spec.rb

```diff
@@ -0,0 +1,336 @@
+        expect_log(params, [2], ['$a == 2', '$b == 2'])
+      end
+
+      it 'will not be evaluated when a value is given' do
+        expect_log(params, [2, 5], ['$a == 2', '$b == 5'])
+      end
+    end
+
+    context 'that references a variable to the right' do
+      let!(:params) { [ <<-SOURCE, 3 ]
+      $a = 10,
+      $b = $c,
+      $c = 20
+      SOURCE
+      }
+
+      it 'fails when the reference is evaluated' do
+        expect_fail(params, [1], /default expression for \$b tries to illegally access not yet evaluated \$c/)
+      end
+
+      it 'does not fail when a value is given for the culprit parameter' do
+        expect_log(params, [1,2], ['$a == 1', '$b == 2', '$c == 20'])
+      end
+
+      it 'does not fail when all values are given' do
+        expect_log(params, [1,2,3], ['$a == 1', '$b == 2', '$c == 3'])
+      end
+    end
+
+    context 'with regular expressions' do
+      it 'evaluates unset match scope parameters to undef' do
+        expect_log([<<-SOURCE, 2], [], ['$a == ', '$b == '])
+        $a = $0,
+        $b = $1
+        SOURCE
+      end
+
+      it 'does not leak match variables from one expression to the next' do
+        expect_log([<<-SOURCE, 2], [], ['$a == [true, h, ello]', '$b == '])
+        $a = ['hello' =~ /(h)(.*)/, $1, $2],
+        $b = $1
+        SOURCE
+      end
+
+      it 'can evaluate expressions in separate match scopes' do
+        expect_log([<<-SOURCE, 3], [], ['$a == [true, h, ell, o]', '$b == [true, h, i, ]', '$c == '])
+        $a = ['hello' =~ /(h)(.*)(o)/, $1, $2, $3],
+        $b = ['hi' =~ /(h)(.*)/, $1, $2, $3],
+        $c = $1
+        SOURCE
+      end
+
+      it 'can have nexted match expressions' do
+        expect_log([<<-SOURCE, 2], [], ['$a == [true, h, oo, h, i]', '$b == '] )
+        $a = ['hi' =~ /(h)(.*)/, $1, if'foo' =~ /f(oo)/ { $1 }, $1, $2],
+        $b = $0
+        SOURCE
+      end
+
+      it 'can not see match scope from calling scope', :if => call_type == 'Function' do
+        expect_log([<<-SOURCE, <<-BODY], <<-CALL, ['$a == '])
+        $a = $0
+        SOURCE
+        {
+          notice("\\$a == ${a}")
+        }
+        function caller() {
+          example()
+        }
+        BODY
+        $tmp = 'foo' =~ /(f)(o)(o)/
+        caller()
+        CALL
+      end
+
+      context 'matches in calling scope', :if => call_type == 'EPP' do
+        it 'are available when using inlined epp' do
+          # Note that CODE is evaluated before the EPP is evaluated
+          #
+          expect_log([<<-SOURCE, <<-BODY], [], ['$ax == true', '$bx == foo'], <<-CODE, true)
+          $a = $tmp,
+          $b = $0
+          SOURCE
+          <% called_from_template($a, $b) %>
+          BODY
+          function called_from_template($ax, $bx) {
+            notice("\\$ax == $ax")
+            notice("\\$bx == $bx")
+          }
+          $tmp = 'foo' =~ /(f)(o)(o)/
+          CODE
+        end
+
```

Comment:
Additional test is needed: since match variables are available in the parameter scope, a test is needed to ensure that they are not available in the epp body (this because inline_epp that specifies parameters hides the calling scopes variables from the body (but they are available in the parameter scope).